## Warm-up Question 1

#### In this activity, you don't need to rerun the code, since the CAE Jupyter system doesn't allow us to install the required package nltk. We have already run the code locally and are keeping the results below. You can finish this warm-up question 1 by just reading through the code.

#### If you are using your local Jupyter notebook, you can run the first cell to install the package

In [2]:
# Install a pip package in the current Jupyter kernel
import sys
!{sys.executable} -m pip install nltk

In [1]:
# regex for removing punctuation
import re
# nltk: useful text processing library 
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
from collections import Counter
def get_part_of_speech(word):
  probable_part_of_speech = wordnet.synsets(word)
  pos_counts = Counter()
  pos_counts["n"] = len(  [ item for item in probable_part_of_speech if item.pos()=="n"]  )
  pos_counts["v"] = len(  [ item for item in probable_part_of_speech if item.pos()=="v"]  )
  pos_counts["a"] = len(  [ item for item in probable_part_of_speech if item.pos()=="a"]  )
  pos_counts["r"] = len(  [ item for item in probable_part_of_speech if item.pos()=="r"]  )
  
  most_likely_part_of_speech = pos_counts.most_common(1)[0][0]
  return most_likely_part_of_speech

In [2]:
## fly, flies, flying
text = "fly, flies, flying"

cleaned = re.sub('\W+', ' ', text) #removing punctuation
tokenized = word_tokenize(cleaned) #breaking text into individual words

stemmer = PorterStemmer() 
stemmed = [stemmer.stem(token) for token in tokenized] #Stemming each word

lemmatizer = WordNetLemmatizer()
lemmatized = [lemmatizer.lemmatize(token, get_part_of_speech(token)) for token in tokenized] #Lemmatizating each word

print("Stemmed text:")
print(stemmed)

print("\nLemmatized text:")
print(lemmatized)

Stemmed text:
['fli', 'fli', 'fli']

Lemmatized text:
['fly', 'fly', 'fly']


In [3]:
### organize, organizes, organizing
text = "organize, organizes, organizing"

cleaned = re.sub('\W+', ' ', text) #removing punctuation
tokenized = word_tokenize(cleaned) #breaking text into individual words

stemmer = PorterStemmer() 
stemmed = [stemmer.stem(token) for token in tokenized] #Stemming each word

lemmatizer = WordNetLemmatizer()
lemmatized = [lemmatizer.lemmatize(token, get_part_of_speech(token)) for token in tokenized] #Lemmatizating each word

print("Stemmed text:")
print(stemmed)

print("\nLemmatized text:")
print(lemmatized)

Stemmed text:
['organ', 'organ', 'organ']

Lemmatized text:
['organize', 'organize', 'organize']


In [4]:
### universe, university
text = "universe, university"

cleaned = re.sub('\W+', ' ', text) #removing punctuation
tokenized = word_tokenize(cleaned) #breaking text into individual words

stemmer = PorterStemmer() 
stemmed = [stemmer.stem(token) for token in tokenized] #Stemming each word

lemmatizer = WordNetLemmatizer()
lemmatized = [lemmatizer.lemmatize(token, get_part_of_speech(token)) for token in tokenized] #Lemmatizating each word

print("Stemmed text:")
print(stemmed)

print("\nLemmatized text:")
print(lemmatized)

Stemmed text:
['univers', 'univers']

Lemmatized text:
['universe', 'university']


##### Did you find anything interesting? Did the meaning of the individual word change a lot after lemmatization and stemming?

## Warm-up Question 2

#### This activity uses 3 three tweets from Donald Trump's Twiiter. They are tokenized beforehand since the required package is not installed.
#### docA = "I never said the pandemic was a Hoax! Who would say such a thing? I said that the Do Nothing Democrats, together with their Mainstream Media partners, are the Hoax. They have been called out & embarrassed on this, even admitting they were wrong, but continue to spread the lie!"
##### docB = "The people that know me and know the history of our Country say that I am  the hardest working President in history. I don’t know about that, but I am a hard worker and have probably gotten more done in the first 3 1/2 years than any President in history. The Fake News hates it!"
##### docC = "Does anybody get the meaning of what a so-called Noble (not Nobel) Prize is, especially as it pertains to Reporters and Journalists? Noble is defined as, “having or showing fine personal qualities or high moral principles and ideals.” Does sarcasm ever work?"


In [5]:
bowA = ['I', 'never', 'said', 'the', 'pandemic', 'was', 'a', 'Hoax', 'Who', 'would', 'say', 'such', 'a', 'thing', 'I', 'said', 'that', 'the', 'Do', 'Nothing', 'Democrats', 'together', 'with', 'their', 'Mainstream', 'Media', 'partners', 'are', 'the', 'Hoax', 'They', 'have', 'been', 'called', 'out', 'embarrassed', 'on', 'this', 'even', 'admitting', 'they', 'were', 'wrong', 'but', 'continue', 'to', 'spread', 'the', 'lie']
bowB = ['The', 'people', 'that', 'know', 'me', 'and', 'know', 'the', 'history', 'of', 'our', 'Country', 'say', 'that', 'I', 'am', 'the', 'hardest', 'working', 'President', 'in', 'history', 'I', 'don', 't', 'know', 'about', 'that', 'but', 'I', 'am', 'a', 'hard', 'worker', 'and', 'have', 'probably', 'gotten', 'more', 'done', 'in', 'the', 'first', '3', '1', '2', 'years', 'than', 'any', 'President', 'in', 'history', 'The', 'Fake', 'News', 'hates', 'it']
bowC = ['Does', 'anybody', 'get', 'the', 'meaning', 'of', 'what', 'a', 'so', 'called', 'Noble', 'not', 'Nobel', 'Prize', 'is', 'especially', 'as', 'it', 'pertains', 'to', 'Reporters', 'and', 'Journalists', 'Noble', 'is', 'defined', 'as', 'having', 'or', 'showing', 'fine', 'personal', 'qualities', 'or', 'high', 'moral', 'principles', 'and', 'ideals', 'Does', 'sarcasm', 'ever', 'work']
#bow stands for bag of words
print("Document A: "+str(bowA))
print("\nDocument B: "+str(bowB))
print("\nDocument C: "+str(bowC))

Document A: ['I', 'never', 'said', 'the', 'pandemic', 'was', 'a', 'Hoax', 'Who', 'would', 'say', 'such', 'a', 'thing', 'I', 'said', 'that', 'the', 'Do', 'Nothing', 'Democrats', 'together', 'with', 'their', 'Mainstream', 'Media', 'partners', 'are', 'the', 'Hoax', 'They', 'have', 'been', 'called', 'out', 'embarrassed', 'on', 'this', 'even', 'admitting', 'they', 'were', 'wrong', 'but', 'continue', 'to', 'spread', 'the', 'lie']

Document B: ['The', 'people', 'that', 'know', 'me', 'and', 'know', 'the', 'history', 'of', 'our', 'Country', 'say', 'that', 'I', 'am', 'the', 'hardest', 'working', 'President', 'in', 'history', 'I', 'don', 't', 'know', 'about', 'that', 'but', 'I', 'am', 'a', 'hard', 'worker', 'and', 'have', 'probably', 'gotten', 'more', 'done', 'in', 'the', 'first', '3', '1', '2', 'years', 'than', 'any', 'President', 'in', 'history', 'The', 'Fake', 'News', 'hates', 'it']

Document C: ['Does', 'anybody', 'get', 'the', 'meaning', 'of', 'what', 'a', 'so', 'called', 'Noble', 'not', 'No

In [6]:
### 1) Run the following code to construct a term-frequency matrix in a simple way
#First create a word set that contains all words in all documents
wordSet = set(bowA).union(set(bowB)).union(set(bowC))
#Second create dictionaries to keep my word counts.
wordDictA = dict.fromkeys(wordSet,0)
wordDictB = dict.fromkeys(wordSet,0)
wordDictC = dict.fromkeys(wordSet,0)
#Then count the words in each document
for word in bowA:
    wordDictA[word] += 1
for word in bowB:
    wordDictB[word] += 1
for word in bowC:
    wordDictC[word] += 1    

In [7]:
import pandas as pd
pd.DataFrame([wordDictA, wordDictB, wordDictC])

1  2  3  Country  Democrats  Do  Does  Fake  Hoax  I  ...  was  were  what  \
0  0  0  0        0          1   1     0     0     2  2  ...    1     1     0   
1  1  1  1        1          0   0     0     1     0  3  ...    0     0     0   
2  0  0  0        0          0   0     2     0     0  0  ...    0     0     1   

   with  work  worker  working  would  wrong  years  
0     1     0       0        0      1      1      0  
1     0     0       1        1      0      0      1  
2     0     1       0        0      0      0      0  

[3 rows x 106 columns]

In [8]:
### 2) Complete the following code to construct the matrix using TF-IDF
#First define a function to compute the term frequency TF
def computeTF(wordDict, bow):
    tfDict = {}
    bowCount = len(bow)
    for word, count in wordDict.items():
        tfDict[word] = count / float(bowCount)
    return tfDict

In [9]:
tfBowA = computeTF(wordDictA, bowA)
tfBowB = computeTF(wordDictB, bowB)
tfBowC = computeTF(wordDictC, bowC)

The cell below has an unfinished equation that needs to be completed

In [10]:
#Second define a function to compute the inverse document frequency IDF
def computeIDF(docList):
    import math
    idfDict = {}
    N = len(docList)
    
    #counts the number of documents that contain a word w
    idfDict = dict.fromkeys(docList[0].keys(), 0)
    for doc in docList:
        for word, val in doc.items():
            if val > 0:
                idfDict[word] += 1
    
    #divide N by denominator above, take the log of that
    for word, val in idfDict.items():
        #Complete the code here
        idfDict[word] = math.log(N / float(val))
    
    return idfDict

In [11]:
idfs = computeIDF([wordDictA, wordDictB, wordDictC])

In [12]:
#Lastly, compute TF-IDF
def computeTFIDF(tfBow, idfs):
    tfidf = {}
    for word, val in tfBow.items():
        tfidf[word] = val * idfs[word]
    return tfidf

In [13]:
tfidfBowA = computeTFIDF(tfBowA, idfs)
tfidfBowB = computeTFIDF(tfBowB, idfs)
tfidfBowC = computeTFIDF(tfBowC, idfs)
pd.DataFrame([tfidfBowA, tfidfBowB, tfidfBowC])

1         2         3   Country  Democrats        Do      Does  \
0  0.000000  0.000000  0.000000  0.000000   0.022421  0.022421  0.000000   
1  0.019274  0.019274  0.019274  0.019274   0.000000  0.000000  0.000000   
2  0.000000  0.000000  0.000000  0.000000   0.000000  0.000000  0.051098   

       Fake      Hoax        I  ...       was      were      what      with  \
0  0.000000  0.044841  0.01655  ...  0.022421  0.022421  0.000000  0.022421   
1  0.019274  0.000000  0.02134  ...  0.000000  0.000000  0.000000  0.000000   
2  0.000000  0.000000  0.00000  ...  0.000000  0.000000  0.025549  0.000000   

       work    worker   working     would     wrong     years  
0  0.000000  0.000000  0.000000  0.022421  0.022421  0.000000  
1  0.000000  0.019274  0.019274  0.000000  0.000000  0.019274  
2  0.025549  0.000000  0.000000  0.000000  0.000000  0.000000  

[3 rows x 106 columns]

### Warm-up Question 3

In [16]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD
import pandas as pd

In [17]:
#In this question, we will use a dataset that contains posts from a data science forum
postDocs = pd.read_csv("post_parsed.csv")
postDocs = postDocs["text"].tolist()

In [18]:
#Create own list of stopwords
nltk_stopword = ["i", "me", "my", "myself", "we", "our", "ours", "ourselves", "you", "your", "yours", "yourself", "yourselves", "he", "him", "his", "himself", "she", "her", "hers", "herself", "it", "its", "itself", "they", "them", "their", "theirs", "themselves", "what", "which", "who", "whom", "this", "that", "these", "those", "am", "is", "are", "was", "were", "be", "been", "being", "have", "has", "had", "having", "do", "does", "did", "doing", "a", "an", "the", "and", "but", "if", "or", "because", "as", "until", "while", "of", "at", "by", "for", "with", "about", "against", "between", "into", "through", "during", "before", "after", "above", "below", "to", "from", "up", "down", "in", "out", "on", "off", "over", "under", "again", "further", "then", "once", "here", "there", "when", "where", "why", "how", "all", "any", "both", "each", "few", "more", "most", "other", "some", "such", "no", "nor", "not", "only", "own", "same", "so", "than", "too", "very", "s", "t", "can", "will", "just", "don", "should", "now"]
stopset = set(nltk_stopword)
stopset.update(['lt','p','/p','br','amp','quot','field','font','normal','span','0px','rgb','style','51', 
                'spacing','text','helvetica','size','family', 'space', 'arial', 'height', 'indent', 'letter'
                'line','none','sans','serif','transform','line','variant','weight','times', 'new','strong', 'video', 'title'
                'white','word','letter', 'roman','0pt','16','color','12','14','21', 'neue', 'apple', 'class',  ])

In [20]:
#Use scikit-learn's TF-IDF vectorizer to take my corpus and convert each document into a sparse matrix of TFIDF Features
#Before vectorizing
postDocs[0]

'<p>data science is about analyzing relevant data to obtain patterns of information in order to help achieve a goal. the main focus of the data analysis is the goal rather then the methodology on how it will achieved. this allows for creative thinking and allowing for the optimal solution or model to be found wihtout the constraint of a specific methodology.</p>'

In [21]:
vectorizer = TfidfVectorizer(stop_words=stopset, use_idf=True, ngram_range=(1, 3)) #ngram_range allows the vectorizer to tokenize not only a single word but also bigram and trigram
X = vectorizer.fit_transform(postDocs)
X

<27x3451 sparse matrix of type '<class 'numpy.float64'>'
	with 4091 stored elements in Compressed Sparse Row format>

The vectorized X is a sparse matrix with lots of 0 elements, which means there are lots of TF-IDF scores equal to 0.To save some memory of the machine, the matrix only stores the elements that have a score and washes out the 0 elements.

In [22]:
X[0]

<1x3451 sparse matrix of type '<class 'numpy.float64'>'
	with 89 stored elements in Compressed Sparse Row format>

In [23]:
#After vectorizing
print(X[0])

  (0, 654)	0.08989762673289421
  (0, 2484)	0.031055663877907407
  (0, 160)	0.06750601481982778
  (0, 2413)	0.0882767769894731
  (0, 2042)	0.10904753915911843
  (0, 2156)	0.09689742217847831
  (0, 1588)	0.052957348020145734
  (0, 2087)	0.07612666000883297
  (0, 1472)	0.07612666000883297
  (0, 47)	0.09689742217847831
  (0, 1389)	0.1765535539789462
  (0, 1817)	0.10904753915911843
  (0, 1261)	0.0882767769894731
  (0, 143)	0.07150739991692207
  (0, 2378)	0.10904753915911843
  (0, 1918)	0.21809507831823685
  (0, 52)	0.10904753915911843
  (0, 108)	0.10904753915911843
  (0, 630)	0.09689742217847831
  (0, 2980)	0.09689742217847831
  (0, 105)	0.10904753915911843
  (0, 2081)	0.10904753915911843
  (0, 2755)	0.09689742217847831
  (0, 1946)	0.07612666000883297
  (0, 1295)	0.0882767769894731
  :	:
  (0, 2044)	0.10904753915911843
  (0, 2158)	0.10904753915911843
  (0, 1600)	0.10904753915911843
  (0, 2093)	0.10904753915911843
  (0, 1474)	0.10904753915911843
  (0, 49)	0.10904753915911843
  (0, 1391)	0.10

Even though 3451 elements are in the original matrix, the vectorizer only scores 89 of them. For example, the 654th word in the first document has a TF-IDF score of 0.08989762673289421 

In [27]:
##Latent Semantic Analysis
X.shape

(27, 3451)

In [28]:
lsa = TruncatedSVD(n_components=27, n_iter=100) # n_components specifies the number of topics.
lsa.fit(X)

TruncatedSVD(algorithm='randomized', n_components=27, n_iter=100,
             random_state=None, tol=0.0)

### LSA

$$X \approx USV^{T}$$

Input: X, a matrix where m is the number of documents, and n is the number of terms.

U will be a m x k matrix. The rows will be documents and the columns will be 'topics'

S will be a k x k diagnal matrix. The elements will be the amount of variation captured from each topic.

V will be a n x k matrix. The rows will be terms and the columns will be topics.  

In [29]:
#This is the first row for V
lsa.components_[0] #lsa.components_ is designed to represent V, which is the term by topic matrix

array([0.00476404, 0.00476404, 0.00476404, ..., 0.00476404, 0.00476404,
       0.00476404])

In [30]:
#Attach the terms to the components and print the first 10 terms that make up the topics
terms = vectorizer.get_feature_names()
for i, comp in enumerate(lsa.components_): 
    termsInComp = zip (terms,comp)
    sortedTerms =  sorted(termsInComp, key=lambda x: x[1], reverse=True) [:10]
    print("Topic %d:" % i )
    for term in sortedTerms:
        print(term[0])
    print (" ")

Topic 0:
data
large amounts
large amounts data
amounts
amounts data
different
large
science
used
data science
 
Topic 1:
large amounts
large amounts data
amounts
amounts data
used
large
according data
according data science
amounts data procedures聽could
amounts data since
 
Topic 2:
white
converted
white converted
white white
big
big data
hello
and聽 white
and聽 white converted
big data and聽
 
Topic 3:
make
decisions
problem
make better
data science analyzing
science analyzing
white
better
better decisions
make better decisions
 
Topic 4:
goal
data science analyzing
science analyzing
achieve
solution
methodology
relevant
relevant data
information
finding
 
Topic 5:
business
goal
methods
competitive edge
edge
especially
achieve
solution
perspective
analyzing
 
Topic 6:
art
answer
using
relevant
relevant data
part
using relevant
using relevant data
good
able learn
 
Topic 7:
data science
science
ultimately
understanding
predict
insights
people
information
digital
trends
 
Topic 8:
question

#### After printing out the topics and their related terms, could you summarize the topics using your own words?   

#### Try to use a sentence or a phrase to summarize topic 3, 4 and 7. For example, topic 0 could be summed up as “data science uses big data”.